File: unityapp1/Assets/Scripts/Game/Direction.cs
namespace Sokoban.Game
{
    public enum Direction
    {
        Up, Down, Left, Right
    }

    public static class DirectionUtil
    {
        public static void ToDelta(Direction d, out int dx, out int dy)
        {
            dx = dy = 0;
            switch (d)
            {
                case Direction.Up: dy = -1; break;
                case Direction.Down: dy = 1; break;
                case Direction.Left: dx = -1; break;
                case Direction.Right: dx = 1; break;
            }
        }

        public static bool TryParse(char c, out Direction d)
        {
            switch (char.ToUpperInvariant(c))
            {
                case 'U': d = Direction.Up; return true;
                case 'D': d = Direction.Down; return true;
                case 'L': d = Direction.Left; return true;
                case 'R': d = Direction.Right; return true;
                default: d = Direction.Up; return false;
            }
        }
    }
}

File: androidapp1/app/src/main/java/com/whj/sokoban/game/Direction.kt
package com.whj.sokoban.game

enum class Direction(val dx: Int, val dy: Int, val code: Char) {
    UP(0, -1, 'U'),
    DOWN(0, 1, 'D'),
    LEFT(-1, 0, 'L'),
    RIGHT(1, 0, 'R');

    companion object {
        fun fromCode(ch: Char): Direction? = when (ch.uppercaseChar()) {
            'U' -> UP
            'D' -> DOWN
            'L' -> LEFT
            'R' -> RIGHT
            else -> null
        }

        fun fromDelta(dx: Int, dy: Int): Direction? = entries.find { it.dx == dx && it.dy == dy }
    }
}
